src/ghga_service_commons/utils/multinode_storage.py

```diff
@@ -0,0 +1,53 @@
+# Copyright 2021 - 2023 Universität Tübingen, DKFZ, EMBL, and Universität zu Köln
+# for the German Human Genome-Phenome Archive (GHGA)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+"""Configuration for multiple object storage nodes"""
+
+from hexkit.providers.s3 import S3Config, S3ObjectStorage
+from pydantic_settings import BaseSettings
+
+
+class ObjectStorageNodeConfig(BaseSettings):
+    """Configuration for one specific object storage node"""
+
+    bucket: str
+    credentials: S3Config
+
+
+class ObjectStorageConfig(BaseSettings):
+    """Configuration for all available object storage nodes indexed by location label"""
+
+    object_storages: dict[str, ObjectStorageNodeConfig]
+
+
+class ObjectStorages:
```

Comment:
Not sure how we can apply this without breaking the hexagonal structure. I.e. there is no abstract version of ObjectStorage. (That's why the current class should rather be called S3ObjectStorages, however, this does not solve the higher problem.)

Reply:
I'll have to try something slightly different.
The major problem here was accessing the config inside the dependency injector solution to directly parse the inner Config. Might work now with out own DI solution

Reply:
What about something more lazily evaluated like this?
```
class ObjectStorages(ABC):
    @abstractmethod
    def for_alias(self, endpoint_alias: str) -> tuple[str, ObjectStorageProtocol]:
        """Get bucket name and config for a specific alias"""


class S3ObjectStorages(ObjectStorages):
    def __init__(self, *, config: S3ObjectStorageConfig):
        self._config = config

    def for_alias(self, endpoint_alias: str) -> tuple[str, S3ObjectStorage]:
        """Get bucket name and config for a specific alias"""
        node_config = self._config.object_storages[endpoint_alias]
        return node_config.bucket, S3ObjectStorage(config=node_config.credentials)
```

Reply:
Looks good, however, we still need to find a solution to have one S3ObjectStorage instance shared for multiple buckets so that we can perform a copy operation.